plugins/thunderbolt/fu-plugin-thunderbolt.c

```diff
@@ -0,0 +1,736 @@
+	if (uuid == NULL)
+		return;
+
+	dev = g_hash_table_lookup (up_data->changes, uuid);
+	if (dev == NULL) {
+		/* a previously unknown device, add it via
+		 * the normal way */
+		fu_plugin_thunderbolt_add (plugin, device);
+		return;
+	}
+
+	/* maybe the device path has changed, lets make sure
+	 * it is correct */
+	path = g_udev_device_get_sysfs_path (device);
+	fu_device_set_metadata (dev, "sysfs-path", path);
+
+	/* make sure the version is correct, might have changed
+	 * after update. */
+	version = g_udev_device_get_sysfs_attr (device, "nvm_version");
+	fu_device_set_version (dev, version);
+
+	id = fu_plugin_thunderbolt_gen_id (device);
+	fu_plugin_cache_add (plugin, id, dev);
+
+	g_hash_table_remove (up_data->changes, uuid);
+
+	/* check if this device is the target*/
+	if (g_str_equal (uuid, up_data->target_uuid)) {
+		up_data->have_device = TRUE;
+		g_debug ("target (%s) re-appeared", uuid);
+		g_main_loop_quit (up_data->mainloop);
+	}
+}
+
+static void
+on_wait_for_device_removed (FuPlugin    *plugin,
+			    GUdevDevice *device,
+			    UpdateData *up_data)
+{
+	g_autofree gchar *id = NULL;
+	FuDevice  *dev;
+	const gchar *uuid;
+
+	id = fu_plugin_thunderbolt_gen_id (device);
+	dev = fu_plugin_cache_lookup (plugin, id);
+
+	if (dev == NULL)
+		return;
+
+	fu_plugin_cache_remove (plugin, id);
+	uuid = fu_device_get_id (dev);
+	g_hash_table_insert (up_data->changes,
+			     (gpointer) uuid,
+			     g_object_ref (dev));
+
+	/* check if this device is the target*/
+	if (g_str_equal (uuid, up_data->target_uuid)) {
+		up_data->have_device = FALSE;
+		g_debug ("target (%s) disappeared",
+			 uuid);
+	}
+}
+
+static void
+on_wait_for_device_notify (FuPlugin    *plugin,
+			   GUdevDevice *device,
+			   const char  *action,
+			   gpointer    user_data)
+{
+	UpdateData *up_data = (UpdateData *) user_data;
+
+	if (g_str_equal (action, "add")) {
+		on_wait_for_device_added (plugin, device, up_data);
+	} else if (g_str_equal (action, "remove")) {
+		on_wait_for_device_removed (plugin, device, up_data);
+	} else if (g_str_equal (action, "change")) {
+		fu_plugin_thunderbolt_change (plugin, device);
+	}
+}
+
+static void
+remove_leftover_devices (gpointer key,
+			 gpointer value,
+			 gpointer user_data)
+{
+	FuPlugin  *plugin = (FuPlugin *) user_data;
+	FuDevice *dev = (FuDevice *) value;
+	const gchar *syspath = fu_device_get_metadata (dev, "sysfs-path");
+	g_autofree gchar *id = NULL;
+
+	id = fu_plugin_thunderbolt_gen_id_from_syspath (syspath);
+
+	fu_plugin_cache_remove (plugin, id);
+	fu_plugin_device_remove (plugin, dev);
+}
+
+static gboolean
+fu_plugin_thunderbolt_wait_for_device (FuPlugin  *plugin,
+				       FuDevice  *dev,
+				       GError   **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	UpdateData    up_data = { TRUE, };
+	g_autoptr(GMainLoop) mainloop = NULL;
+	g_autoptr(GHashTable) changes = NULL;
+
+	up_data.mainloop = mainloop = g_main_loop_new (NULL, FALSE);
```

Comment:
I'm also wondering if it makes sense to just make it a g_autoptr() and then just use `G_DEFINE_AUTOPTR_CLEANUP_FUNC` -- then you don't have to do this workaround.